services/supervisor/src/lib.rs

```diff
@@ -432,8 +525,148 @@ impl Service for Supervisor {
+        let requests_to_remove = SchemaImpl::new(context.service_data())
+            .pending_migrations
+            .iter()
+            .filter_map(|(_, request)| {
+                if request.deadline_height <= height {
+                    Some(request)
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
+
+        for request in requests_to_remove {
+            let mut schema = SchemaImpl::new(context.service_data());
+            schema.pending_migrations.remove(&request);
+
+            let mut state = schema
+                .migration_states
+                .get(&request)
+                .expect("BUG: State for pending migration request is not stored");
+            if state.state.is_pending() {
+                // If state is marked as pending, change it to failed as well.
+                state.fail(AsyncEventState::Timeout);
+                schema.migration_states.put(&request, state);
+
+                // Then, rollback the migration.
+                drop(schema);
+                context
+                    .supervisor_extensions()
+                    .rollback_migration(request.service.as_ref())?;
+            }
+            log::trace!("Removed outdated migration request {:?}", request);
+        }
+
+        Ok(())
+    }
+
+    /// Goes through incomplete migrations, checking their statuses.
+    fn process_incomplete_migrations(&self, context: &mut AfterCommitContext<'_>) {
+        let service_key = context.service_key();
+
+        // First of all, check all the new migrations and request core to start them.
+        let pending_migrations: Vec<_> = {
+            let schema = SchemaImpl::new(context.service_data());
+            schema
+                .pending_migrations
+                .iter()
+                .filter_map(|(_, request)| {
+                    let state = schema
+                        .migration_states
+                        .get(&request)
+                        .expect("BUG: State for pending migration request is not stored");
+
+                    let confirmed_by_us = schema
+                        .migration_confirmations
+                        .confirmed_by(&request, &service_key);
+
+                    // We are interested in requests that are both pending and not yet confirmed.
+                    // Despite the fact that `migration_confirmations` stores only successful
+                    // outcomes, receiving any failure report will immediately change the state
+                    // to `Failed`.
+                    // Thus if request is `pending` and there is no out signature in
+                    // `migration_confirmations` index, it means than we did not send the
+                    // result report, and should do it once core will provide this result.
+                    if state.state.is_pending() && !confirmed_by_us {
```

Comment:
Actually, it was `inner` in the previous iteration, but then I've renamed it because `inner` won't look adequate in the serialized form. I can just use `serde(rename = "state")` though.